core/src/main/scala/chisel3/VerificationStatement.scala

```diff
@@ -84,15 +122,14 @@ object assert {
   def _applyWithSourceLine(
     cond:    Bool,
     line:    SourceLineInfo,
-    message: Option[String],
-    data:    Bits*
+    message: Option[Printable]
   )(
```

Comment:
This is a binary incompatible change (because this method is injected into user code by our macros). Thus we need to keep around the old version (not deprecated, just marked with a comment to delete in 3.6).

You should be able to overload it though, both versions can have the same name.

Reply:
Agreed!

Reply:
@jackkoenig  tried this 
```def _applyWithSourceLine(
    cond:    Bool,
    line:    SourceLineInfo,
    message: Option[String],
    data:    Bits*
  )(
    implicit sourceInfo: SourceInfo,
    compileOptions:      CompileOptions
  ): Assert = {
    val id = new Assert()
    when(!Module.reset.asBool()) {
      failureMessage("Assertion", line, cond,message.map(Printable.pack(_,data : _*))) 
      Builder.pushCommand(Verification(id, Formal.Assert, sourceInfo, Module.clock.ref, cond.ref, ""))
    }
    id
  }
```
And now compile fails telling there is ambiguity - I understand why but yet to figure out a solution. Still digging - but putting it out here. 

```
ambiguous reference to overloaded definition,
[error] both method _applyWithSourceLine in object assert of type (cond: chisel3.Bool, line: chisel3.VerificationStatement.SourceLineInfo, message: Option[chisel3.Printable])(implicit sourceInfo: chisel3.internal.sourceinfo.SourceInfo, implicit compileOptions: chisel3.CompileOptions)chisel3.assert.Assert
[error] and  method _applyWithSourceLine in object assert of type (cond: chisel3.Bool, line: chisel3.VerificationStatement.SourceLineInfo, message: Option[String], data: chisel3.Bits*)(implicit sourceInfo: chisel3.internal.sourceinfo.SourceInfo, implicit compileOptions: chisel3.CompileOptions)chisel3.assert.Assert
[error] match argument types (chisel3.Bool,(String, Int, String),None.type)
[error]   assert((a.andR) === (s"%${w}s".format(BigInt(_a).toString(2)).foldLeft(true)(_ && _ == '1')).asBool) 
```

Reply:
At worse I need to overload the failureMessage! Initially my intention was to reduce duplication as much as possible -since I was not aware of the additional constraint of binary compatibility. 

@jackkoenig  feel free to suggest any other way out of this - otherwise I will go ahead with the obvious solution

Reply:
Ok I take back what I said - basically I need to just explicitly change the name of my new method (append Printable or something) and we should be good! Trying ....

Reply:
Still getting same error - need to take a break from screen and get back to this - not thinking clearly I guess!

Reply:
Ok - that solution of renaming my new _appltSourceLinePrintable  to append with Printable worked - so nothing within this file references the old _applySourceLine method. 

```
/** This will be removed in Chisel 3.6 in favor of the Printable version */
  def _applyWithSourceLine(
    cond:    Bool,
    line:    SourceLineInfo,
    message: Option[String],
    data:    Bits*
  )(
    implicit sourceInfo: SourceInfo,
    compileOptions:      CompileOptions
  ): Assume = {
    val id = new Assume()
    when(!Module.reset.asBool()) {
      failureMessage("Assumption", line, cond, message.map(Printable.pack(_,data: _*)))
      Builder.pushCommand(Verification(id, Formal.Assume, sourceInfo, Module.clock.ref, cond.ref, ""))
    }
    id
  }

  /** Used by our macros. Do not call directly! */
  def _applyWithSourceLinePrintable(
    cond:    Bool,
    line:    SourceLineInfo,
    message: Option[Printable]
  )(
    implicit sourceInfo: SourceInfo,
    compileOptions:      CompileOptions
  ): Assume = {
    val id = new Assume()
    when(!Module.reset.asBool()) {
      failureMessage("Assumption", line, cond, message)
      Builder.pushCommand(Verification(id, Formal.Assume, sourceInfo, Module.clock.ref, cond.ref, ""))
    }
    id
  }


```
Still curious why the previous version did not work.

Reply:
oh so basically the optionality of data is what causes the ambiguity - despite two different Option types - None is common to both. 

```
def _applyWithSourceLine(
    cond:    Bool,
    line:    SourceLineInfo,
    message: Option[String],
    data:    Bits*
  )(
```

Reply:
@jackkoenig  sadly looks like we might have to live with the uglier solution since using the same name does not work (feel free to suggest alternate solutions)

Reply:
Went ahead with this change to rename the new version with Printable appended and keeping the old one.

Reply:
Yeah we just have to use 2 names. Due to type erasure, `Option[String]` and `Option[Printable]` are actually the same type at runtime so it does make sense that we can't overload here.